app/controllers/actions/IdentifierAction.scala

```diff
@@ -45,12 +45,16 @@ class AuthenticatedIdentifierAction @Inject()(
     implicit val hc = HeaderCarrierConverter.fromRequestAndSession(request, request.session)
 
     authorised(AuthProviders(AuthProvider.Verify) or (AffinityGroup.Individual and ConfidenceLevel.L200))
-      .retrieve(OptionalRetrieval("internalId", Reads.StringReads) and OptionalRetrieval("nino", Reads.StringReads)) {
-        x =>
-          val internalId = x.a.getOrElse(throw new UnauthorizedException("Unable to retrieve internalId"))
-          val nino = x.b.getOrElse(throw new UnauthorizedException("Unable to retrieve nino"))
-
-          block(IdentifierRequest(request, internalId, nino))
+      .retrieve(internalId and nino and saUtr) {
+        case mayBeId ~ mayBeNino ~ mayBeSaUtr =>
```

Comment:
I don't know what this code does but I don't think that matters.

Reply:
@jgrath see https://github.com/hmrc/auth-client#readme

> Note the ~ for chaining the values in the pattern. While this syntax is unusual, it is the only cheap way of giving you any number of type-safe results inside the function without introducing a rather heavy-weight library like shapeless. You can combine any number of retrievals with and. A logical or is not available here as it would not make any sense for retrievals. That's all there is to it. If you were one of the unlucky users who had to use one of our older auth libraries which forced you to implement an AuthProvider, a TaxRegime and maybe also a PageVisibilityPredicate and whatnot, I hope you appreciate the simplicity and consistency of the new library. We did not implement a single trait in the examples above, we merely used existing building blocks to express our requirements!